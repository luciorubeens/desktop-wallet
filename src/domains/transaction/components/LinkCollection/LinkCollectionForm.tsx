import { Button } from "app/components/Button";
import { Form, FormField, FormLabel } from "app/components/Form";
import { Input } from "app/components/Input";
import { Select } from "app/components/SelectDropdown";
import React from "react";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";

import { EntityLink } from "./LinkCollection.models";

type Props = {
	selectOptions: { label: string; value: string }[];
	append: (link: EntityLink) => void;
	itemLabel: string;
};

export const LinkCollectionForm = ({ selectOptions, append, itemLabel }: Props) => {
	const { t } = useTranslation();

	// TODO: Validate URL
	const form = useForm();
	const { register, reset } = form;

	const handleSubmit = (link: EntityLink) => {
		reset();
		append(link);
	};

	return (
		<Form
			data-testid="LinkCollectionForm"
			context={form}
			onSubmit={(data: any) => handleSubmit(data as EntityLink)}
			className="grid col-gap-2 row-gap-4"
			style={{
				gridTemplateColumns: "40% 60%",
			}}
		>
			<FormField name="type">
				<FormLabel label={`Add ${itemLabel}`} />
				<Select options={selectOptions} ref={register({ required: true })} />
			</FormField>

			<FormField name="value">
				<FormLabel label={t("COMMON.LINK")} />
				<Input data-testid="LinkCollectionForm__value" ref={register({ required: true })} />
			</FormField>

			<Button data-testid="LinkCollectionForm__submit" className="col-span-2" variant="plain" type="submit">
				{t("TRANSACTION.ADD_LINK")}
			</Button>
		</Form>
	);
};
