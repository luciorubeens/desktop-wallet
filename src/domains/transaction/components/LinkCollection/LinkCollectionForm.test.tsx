import React from "react";
import { act, fireEvent, render, waitFor } from "utils/testing-library";

import { LinkCollectionForm } from "./LinkCollectionForm";

describe("LinkCollectionForm", () => {
	const options = [
		{
			label: "Example",
			value: "example0",
		},
	];

	it("should add link and clear the form after submitting", async () => {
		const append = jest.fn();
		const { getByTestId, getAllByRole, asFragment } = render(
			<LinkCollectionForm selectOptions={options} itemLabel="example" append={append} />,
		);

		act(() => {
			fireEvent.click(getByTestId("select-list__toggle-button"));
		});

		act(() => {
			fireEvent.click(getAllByRole("option")[0]);
		});

		const valueField = getByTestId("LinkCollectionForm__value");

		act(() => {
			fireEvent.change(valueField, {
				target: {
					value: "testing link",
				},
			});
		});

		const submit = getByTestId("LinkCollectionForm__submit");

		act(() => {
			fireEvent.click(submit);
		});

		await waitFor(() => expect(append).toHaveBeenCalledWith({ type: "example0", value: "testing link" }));

		expect(valueField).toHaveValue("");
		expect(asFragment()).toMatchSnapshot();
	});
});
