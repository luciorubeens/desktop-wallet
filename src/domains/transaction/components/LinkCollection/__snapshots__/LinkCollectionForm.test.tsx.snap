// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkCollectionForm should add link and clear the form after submitting 1`] = `
<DocumentFragment>
  <form
    class="grid col-gap-2 row-gap-4"
    data-testid="LinkCollectionForm"
    style="grid-template-columns: 40% 60%;"
  >
    <fieldset
      class="sc-AxirZ izhStr flex flex-col space-y-2"
    >
      <label
        class="flex inline-block mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-neutral-dark"
        data-testid="FormLabel"
        for="type"
      >
        Add example
        <div
          class="w-1 h-1 mt-1 ml-1 rounded-full bg-theme-danger-400"
          data-testid="FormLabel__required"
        />
      </label>
      <div
        class="relative w-full"
      >
        <input
          class="sc-AxiKw ffjERP overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-neutral-300 text-theme-neutral-900 transition-colors duration-200 px-4 py-3 sr-only"
          data-testid="select-list__input"
          name="type"
          readonly=""
          type="text"
          value="example0"
        />
        <div
          class="relative w-full cursor-pointer"
        >
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-0-label downshift-0-toggle-button"
              class="sc-fzpans SmPby  is-selected "
              data-testid="select-list__toggle-button"
              id="downshift-0-toggle-button"
              type="button"
            >
              Example
            </button>
            <ul
              aria-labelledby="downshift-0-label"
              class="sc-fzplWN kxNmJG"
              id="downshift-0-menu"
              role="listbox"
              tabindex="-1"
            />
          </div>
          <div
            class="sc-AxhUy sc-AxheI bYHtKZ w-10 pointer-events-none text-theme-neutral"
          >
            <div
              class="sc-AxhCb dxfUlj"
              height="8"
              width="8"
            >
              <svg>
                arrowdown.svg
              </svg>
            </div>
          </div>
        </div>
      </div>
    </fieldset>
    <fieldset
      class="sc-AxirZ izhStr flex flex-col space-y-2"
    >
      <label
        class="flex inline-block mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-neutral-dark"
        data-testid="FormLabel"
        for="value"
      >
        Link
        <div
          class="w-1 h-1 mt-1 ml-1 rounded-full bg-theme-danger-400"
          data-testid="FormLabel__required"
        />
      </label>
      <input
        class="sc-AxiKw ffjERP overflow-hidden w-full bg-theme-background appearance-none rounded border border-theme-neutral-300 text-theme-neutral-900 transition-colors duration-200 px-4 py-3 "
        data-testid="LinkCollectionForm__value"
        name="value"
      />
    </fieldset>
    <button
      class="sc-AxjAm cyswVS col-span-2"
      color="primary"
      data-testid="LinkCollectionForm__submit"
      type="submit"
    >
      Add Link
    </button>
  </form>
</DocumentFragment>
`;
